File: python/partition.py
# Implementations of partition algorithms

def is_partitioned(iterable, func):
    pass_partition_point = False
    for i in iterable :
        if not pass_partition_point :
            if not func(i) :
                pass_partition_point = True
        else :
            if func(i) :
                return False
    return True

def partition_point(iterable, func):
    for i, v in enumerate(iterable) :
        if not func(v) :
            return i

def partition(seq, func):
    pivot = partition_point(seq, func)
    for i in range(pivot, len(seq)) :
        if func(seq[i]) :
            seq[pivot], seq[i] = seq[i], seq[pivot]
            pivot += 1

def partition_copy(seq, func):
    res = []
    remain = []
    for i in seq :
        if func(i) : res.append(i)
        else : remain.append(i)
    res.extend(remain)
    return res

def stable_partition(seq, func):
    for i in range(0, len(seq)) :
        if not func(seq[i]) :
            v = seq[i]
            seq.remove(v)
            seq.append(v)

File: ruby/partition.rb
# Implementations of partition algorithms
module Partition

def is_partitioned(iterable)
  pass_parition_point = false
  iterable.each do |v|
    unless pass_parition_point
      pass_parition_point = true unless yield v
    else
      return false if yield v
    end
  end
  true
end

def partition_point(iterable)
  iterable.each_with_index{|v, i| return i unless yield v}
end

def partition(seq, &block)
  pivot = partition_point(seq, &block)
  (pivot...seq.size).each do |i|
    if yield seq[i]
      seq[pivot], seq[i] = seq[i], seq[pivot]
      pivot += 1
    end
  end
end

def partition_copy(seq)
  true_part, false_part = [], []
  seq.each do |v|
    if yield v
      true_part.push v
    else
      false_part.push v
    end
  end
  true_part + false_part
end

def stable_partition(seq)
  count = seq.size
  seq.each do |i|
    unless yield i
      seq.delete i
      seq.push i
    end
    count -= 1
    return if count == 0
  end
end

end
